ONSdigital/rm-notify-gateway
# Extract different client behaviours into separate classes

# Motivation and Context
The `ConfigurationAwareNotificationClient` was doing 3 different things:
1. Checking some configuration and delegating calls to the notify.gov client 
2. Debug logging (also in `RuntimeExceptionThrowingNotificationClient` and `ExceptionThrowingNotificationClient`)
3. Ignoring calls if the `enabled` property was set to `false`

This was confusing to follow and created a number of challenges when trying to progress with new functionality.

# What has changed
No functionality should be changed.

To address the points above this:
1. Simplifys `ConfigurationAwareNotificationClient`
2. Creates a `LoggingNotificationClient` decorator class which decorates all created client classes to enable the same level of logging previously available
3. Creates a new `NullNotificationClient` which is created instead of `ConfigurationAwareNotificationClient` when the `enabled` property is set to `false`.

# How to test?
This is a refactoring so no functionality should have changed.

**The most important thing to check it that no notifications are send if the `enabled` property is set to false.**
